{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a tutorial on how to model in vivo TF binding using sequnce feature, DNA shape features at flanking regions of motifs and histone modification (HM) features.\n",
    "\n",
    "Description of data:\n",
    "- MYC_pwm.txt : Position weight matrix of MYC binding sites in format of FIMO (Grant et al. 2011) input;\n",
    "- BS.bed : Coordinates of BSs, six columns;\n",
    "- non-BS.bed : Coordinates of non-BSs, six columns;\n",
    "- MYC.txt : Stack non-BSs from non-BS.bed (first column assigned label 0) and BSs from BS.bed (first column assigned label 1), 10 bp upstream and downstream surrounding motifs are displayed;\n",
    "- BS_10_HM.txt : 10 HMs patterns around BSs, 10 HMs are H3K4me2, H3K27ac, H3K27me3, H3K36me3, H3K4me1, H3K4me3, H3K79me2, H4K20me1, H3K9me1, and H3K9me3, numbers are shown in order;\n",
    "- non-BS_10_HM.txt : 10 HMs patterns around non-BSs;\n",
    "- MYC.00000000001 10 HM patterns, same order as MYC.txt. First column shows the label for each sequence, second column shows all 1s, which is intercept term in MLR model;\n",
    "\n",
    "The following will explain how to generate these data and how to model TF binding with L2-regularized MLR of a specific TF, in this case, we use MYC.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we generate data with following steps: generate BS.bed from ChIP-seq data -> generate exactly-matched non-BS.bed from chromatin accessible regions, make sure non-BSs have similar distribution of chromatin accessibility with BSs -> scan 10 HM patterns.\n",
    "BS.bed and non-BS.bed for each TF is in each TF folder. Scripts for generating these files are upon request as many files need to be downloaded beforehand. The command below is for generating HM patterns based on pre-calculated files.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#Generate BS_10_HM.txt.\n",
    "bash ./get_HM_pattern_gm12878.sh MYC ./\n",
    "\n",
    "#Generate non-BS_10_HM.txt\n",
    "bash ./get_HM_pattern_gm12878.sh MYC ./\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Second, calculate significance level of HM pattern differences between BSs and non-BSs. Those numbers are visualized in MYC row in Figure 2A, highlighted by a green rectangle. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#install qvalue library from Bioconductro if encessary: http://bioconductor.org/packages/release/bioc/html/qvalue.html\n",
    "Rscript q_value_from_p_value.R ./ MYC"
   ]
  },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "![description](Figure_2A_MYC.png)"
     ]
    },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Third, generate data as input for L2-regularized MLR models. Data for MLR models are stored under MYC/ folder.\n",
    "- MYC.txt stack non-BSs (first column assigned label 0) and BSs (first column assigned label 1), input of one-hot encoded and DNAshapeR;\n",
    "- MYC.00000000001 data for HM models. Stack 10 HM patterns around non-BSs and BSs, same order as MYC.txt. First column still shows the label for each sequence, second column shows all 1s, which is intercept term in MLR model;\n",
    "- MYC.MGW, MYC.Roll, MYC.ProT, and MYC.HelT DNA shape profiles generated by DNAshapeR software (Chiu et al., 2016).\n",
    "- MYC.10000000000 MGW features at flanking regions; \n",
    "- MYC.00010000000 Roll features at flanking regions; \n",
    "- MYC.00001000000 ProT features at flanking regions; \n",
    "- MYC.00000100000 HelT features at flanking regions; \n",
    "- MYC.00000010000 one-hot encoded data for sequence models; \n",
    "- MYC.00011110000 DNA shape data at flanking regions for sequence+shape models, length of flanking regions is adjustable and here we use 10 bp; \n",
    "- MYC.10011110000 data for sequence+shape model; \n",
    "- MYC.10011110001 data for sequence+shape+HM model.",
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sh ./generate_mlr_data.sh MYC 10 ./ 10\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, run L2-regularized MLR models, sequence+shape model, sequence+shape+HM model. Report model performance in AUPRC. Results are visualized in Figure 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Rscript shuffle_divide.R MYC ./MYC/input 10 feature_list_mlr \n",
    "# use ./train.py to get input arguments.\n",
    "./train.py MYC ./MYC/encoded ./MYC/input ./MYC/output 10 ./train.R ./feature_list_mlr\n",
    "./test.py MYC ./MYC/encoded ./MYC/input ./MYC/output 10 ./predict.R ./feature_list_mlr\n",
    "./summarize_mlr_classification.py MYC ./MYC/output/ ./summarize_mlr_classification_auprc.R ./MYC/result ./feature_list_mlr\n"
   ]
  },
  {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "![description](Figure_3A_MYC.png)"
     ]
    },
    {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "References:\n\n",
    "Chiu T, Comoglio F, Zhou T, Yang L, Paro R, Rohs R. 2016. DNAshapeR : an R / Bioconductor package for DNA shape prediction and feature encoding. Bioinformatics 32: 1211–1213.\n\n",
    "Grant CE, Bailey TL, Noble WS. 2011. FIMO: Scanning for occurrences of a given motif. Bioinformatics 27: 1017–1018.\n"
   ]
  }
  ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
